### PR TITLE
Allows for arbitrary ordering of inheritance in type section #6259

### PR DESCRIPTION
**Pull request by jmgomez**

### CODE REVIEW

File: compiler/semstmts.nim [hunk 1]
```diff
@@ -1371,128 +1371,209 @@ proc checkCovariantParamsUsages(c: PContext; genericType: PType) =
...
+    popOwner(c)
+    # If the right hand side expression was a macro call we replace it with
+    # its evaluated result here so that we don't execute it once again in the
+    # final pass
+    if a[2].kind in nkCallKinds:
+      incl a[2].flags, nfSem # bug #10548
+  if sfExportc in s.flags and s.typ.kind == tyAlias:
+    localError(c.config, name.info, "{.exportc.} not allowed for type aliases")
+
+  if tfBorrowDot in s.typ.flags and s.typ.skipTypes({tyGenericBody}).kind != tyDistinct:
```
Comment (metagn) [inline compiler/semstmts.nim:1458]: When rebasing, this needs to be as in #22072

File: compiler/semstmts.nim [hunk 2]
```diff
@@ -1565,7 +1567,7 @@ proc typeSectionFinalPass(c: PContext, n: PNode) =
         # fix bug #5170, bug #17162, bug #15526: ensure locally scoped types get a unique name:
         if s.typ.kind in {tyEnum, tyRef, tyObject} and not isTopLevel(c):
           incl(s.flags, sfGenSym)
-
+
```
Comment (Araq) [inline compiler/semstmts.nim]:
```suggestion

```

File: compiler/semtypes.nim [hunk 3]
```diff
@@ -875,14 +876,14 @@ proc semObjectNode(c: PContext, n: PNode, prev: PType; flags: TTypeFlags): PType
   # n[0] contains the pragmas (if any). We process these later...
   checkSonsLen(n, 3, c.config)
   if n[1].kind != nkEmpty:
-    realBase = semTypeNode(c, n[1][0], nil)
+    realBase = semTypeNode(c, n[1][0], nil)
```
Comment (Araq) [inline compiler/semtypes.nim]:
```suggestion
    realBase = semTypeNode(c, n[1][0], nil)
```

File: compiler/semtypes.nim [hunk 4]
```diff
@@ -875,14 +876,14 @@ proc semObjectNode(c: PContext, n: PNode, prev: PType; flags: TTypeFlags): PType
...
-    realBase = semTypeNode(c, n[1][0], nil)
+    realBase = semTypeNode(c, n[1][0], nil)    
     base = skipTypesOrNil(realBase, skipPtrs)
     if base.isNil:
       localError(c.config, n.info, "cannot inherit from a type that is not an object type")
     else:
       var concreteBase = skipGenericInvocation(base)
       if concreteBase.kind in {tyObject, tyGenericParam,
-        tyGenericInvocation} and tfFinal notin concreteBase.flags:
+        tyGenericInvocation} and tfFinal notin concreteBase.flags:
```
Comment (Araq) [inline compiler/semtypes.nim]:
```suggestion
        tyGenericInvocation} and tfFinal notin concreteBase.flags:
```

**Pull request merged**